# Trabajo final Discretas 2

## Integrantes: 
* Paola Posada Restrepo, 1001242646
* Alejandro Mesa Gómez, 1017228006

## Trabajo:

1) Utilice Diffie-Hellman para ver cu ́al es la clave de encriptaci ́on que intercambian Roberto y Carlos donde $g = 4$ y $p = 8324923$ y el numero de clave privada de Roberto es $234$ y el de Carlos es $545$.

2) Utilizando RSA encuentre la clave d, dado que la clave pública es $(543510200934057811556261, 6121)$

3) Desencripte el siguiente mensaje $(n = 7387, e = 1357)$:
$$2133\text{ } 429 \text{ }1126$$


# Diffie-Hellman

en este proceso se tiene una clave privada (a o b) y una clave publica (g,p), para encontrar el valor de la clave para encriptar se realiza el siguiente proceso:

Roberto elige un a y calcula $r=g^{a}\text{mod}(p)$ y se lo envia a Carlos
Carlos elige un b y calcula $c=g^{b}\text{mod}(p)$y se lo envia a Roberto

Finalmente, ambos toman el valor recibido y lo elevan por el valor de su clave privada, i.e. $(c)^a \text{mod}(p)$ y $(r)^b \text{mod}(p)$

In [1]:
g=4
p=8324923
a=234 #clave de Roberto (emisor)
b=545 #clave de Carlos (receptor)

r=(g**a)%p
c=(g**b)%p
print("valor enviado por Roberto: ",r)
print()
print("valor enviado por Carlos: ",c)
print()

r_1=(c**a)%p
c_1=(r**b)%p


print("clave obtenida por Roberto: ",r_1)
print()
print("clave obtenida por Carlos: ",c_1)

valor enviado por Roberto:  6481925

valor enviado por Carlos:  1220163

clave obtenida por Roberto:  8318058

clave obtenida por Carlos:  8318058


de los cálculos realizados, se concluye que la clave de encriptación usando Diffie-Hellman entre Roberto y Carlos es: $$8318058$$

# RSA

teniedo la clave publica $(n,e)$ donde $n=pq$ números primos  y $e$ es un número tal que mcd$(e,\varphi (n))=1$ entonces la clave privada será $d$ tal que

$$ ed \equiv 1\text{mod}(\varphi(n)) $$

donde $\varphi(n)=(p-1)(q-1)$, para nuestro problema, se tiene que $n=543510200934057811556261$, usando wolfram aplha (https://bit.ly/2siu1zZ) se obtiene que los valores primos p y q son: $p=652873455583$ y $q=832489353467$

In [2]:
###############################################################
## RUTINAS IMPLEMENTADAS
###############################################################

def phi(p,q): #rutina para calcular el phi de n teniendo los valores primos p y q
    return(p-1)*(q-1)

def egcd(a, b):                #maximo comun divisor a partir del algoritmo de Euclides  extendido
    if a == 0:                 #(tomado de https://en.wikibooks.org/wiki/Algorithm_Implementation/Mathematics/Extended_Euclidean_algorithm)
        return (b, 0, 1)       #retorna g,x,y tal que ax + by = g = mcd(a, b).
    else:
        g, y, x = egcd(b % a, a)
        return (g, x - (b // a) * y, y)

def modinv(a, m): #calculo del modulo inverso a partir del egcd
    g, x, y = egcd(a, m)
    if g != 1:
        raise Exception('modular inverse does not exist')
    else:
        return x % m


###############################################################
## PROCESO PARA RSA
###############################################################

n=543510200934057811556261
e=6121
p=652873455583
q=832489353467


phi_n=phi(p,q)
d=modinv(e,phi_n)

print("")
print("con la clave publica: (",n,",",e,") la clave privada d será: ")
print("d = ",d)



con la clave publica: ( 543510200934057811556261 , 6121 ) la clave privada d será: 
d =  67394909738249058748429


finalmente, se concluye que
$$d = 67394909738249058748429 $$

# Desencriptar

para desencriptar se realiza el siguiente proceso: sea $c$ el texto encriptado, $m$ el texto plano original y $d$ la clave privada obtenida por el método de RSA. luego, para obtener el texto desencriptaco a partir de $c$ se aplica la siguiente congruencia:

$$c^d \equiv m \text{mod}(n)$$

teniendo el siguiente diccionario:

|||||||||||
|---|---|---|---|---|---|---|---|---|---|
|A |B |C |D |E |F |G |H |I |J|
|11 |12 |13 |14 |15 |16 |17 |18 |19 |20|
|K |L |M |N |O |P |Q |R |S |T|
|21| 22| 23 |24| 25 |26 |27| 28 |29 |30|
|U |V| W |X |Y |Z|
|31 |32| 33 |34 |35| 36|




se desencripta el texto recibido: $2133\text{ } 429 \text{ }1126$ 


y luego se separan los números en pares para descubrir el mensaje secreto.
Con la clave publica (n = 7387, e = 1357)

In [3]:
dictionary ={11:"A",12:"B",13:"C",14:"D",15:"E",16:"F",18:"G",19:"I",20:"J",
            21:"K",22:"L", 23:"M",24:"N",25:"O",26:"P",27:"Q",28:"R",29:"S",
             30:"T",31:"U",32:"V",33:"W",34:"X",35:"Y",36:"Z"}

def decrypter (m = []):
    data=[]
    for i in m:
        data.append(dictionary[int(i)])
    return data
import re

def concater(m1,m2,m3):
    return re.findall('..',str(m1))+re.findall('..',str(m2))+re.findall('..',str(m3))



In [4]:
n=7387
e=1357
p=83 #obtenido de la misma forma del punto anterior
q=89

phi_n=phi(p,q)
d=modinv(e,phi_n)

C1=2133
C2=429
C3=1126

m1=C1**d%n
m2=C2**d%n
m3=C3**d%n

print("")
print("texto desencriptado (en sus valores numericos): ",m1,m2,m3)
print("")

m=concater(m1,m2,m3)

text=decrypter(m)

palabra=""

palabra=palabra.join(text)


print ("palabra des encriptada: ",palabra)
print("")


texto desencriptado (en sus valores numericos):  2431 2312 1528

palabra des encriptada:  NUMBER



finalmente, usando los metodos creados en este script, se halla que la palabra secreta es:

$$\text{NUMBER}$$